Use semantic `Text` component for component typography

sam-b-rose opened the pull request:
Closes https://github.com/Shopify/polaris-internal/issues/1483

This PR replaces seeks to reduce the need to semantic CSS custom properties for typography by:

- Replacing font styles with the semantic `Text` equivalent
- Remove typography overrides (i.e., `LegacyCard`)
- Remove semantic typography tokens from `Text`
  - **Why?** Using the semantic tokens doesn't support responsive typography

The goal is to allow the `Text` component to apply semantic type styles rather using CSS custom properties. Using the theme and CSS tokens should be a last resort for styling. Using components to compose styles reduces the surface area of maintenance and the likelihood of implementation bugs.

### WHAT is this pull request doing?

- Adds `base`, `inherit`, and `text-inverse-secondary` tones to `Text` component
- Uses the `Text` component to apply typography styles for Polaris components where possible
- Updates the `@shopify/stylelint-polaris` rule to include the following as a warning in the property disallow list:
  - `font-size`
  - `font-weight`
  - `line-height`
  - `letter-spacing`
- Adds a page on polaris.shopify.com for the `typography/property-disallow-list` rule

#### Components updated

- `BulkActions`
- `CheckableButton`
- `DropZone`
- `FooterHelp`
- `Frame`
- `Toast`
- `InlineError`
- `IndexFilters`
- `IndexTable`
- `Label`
- `MediaCard`
- `LegacyCard`
- `Navigation`
- `Page`
- `Select`
- `SelectAllActoins`
- `SkeletonPage`
- `Tabs`
- `Text` (added `base`, `inherit`, and `text-inverse-secondary` tone)
- `TextField`
- `TopBar/UserMenu`

#### Stylelint rule in-action

https://github.com/Shopify/polaris/assets/11774595/d60c3c77-022d-4799-b2ea-5309a8b96bbb

#### New docs page for Stylelint rule

<img width="822" alt="Screenshot 2024-03-21 at 3 53 01 PM" src="https://github.com/Shopify/polaris/assets/11774595/6d28281d-0efc-4961-8319-aadd6b5967e5">

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#install-dependencies-and-build-workspaces)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

### 🎩 checklist

- [x] Tested a [snapshot](https://github.com/Shopify/polaris/blob/main/documentation/Releasing.md#-snapshot-releases)
- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

alex-page commented on polaris-react/src/components/Filters/components/FiltersBar/FiltersBar.tsx:
```diff
@@ -162,8 +160,6 @@ export function FiltersBar({
 
   const hasOneOrMorePinnedFilters = pinnedFilters.length >= 1;
 
-  const labelVariant = mdDown ? 'bodyMd' : 'bodySm';
```
Are we okay with removing this font size swap at breakpoint?

sam-b-rose replied:
Yes, I think it is currently too large on Mobile when the `bodyMd` is being used. With the responsive text styles, the `bodySm` gets larger now. I'd like to check with @mrcthms though since he has more context on Filters

| Before | After |
| --- | --- |
| <img width="267" alt="21-38-on21u-4d0p7" src="https://github.com/Shopify/polaris/assets/11774595/5514c41c-5ac3-4409-96b7-c18ad532bdcd"> | <img width="241" alt="21-39-k1mel-18oap" src="https://github.com/Shopify/polaris/assets/11774595/86e6956f-9b68-4656-bf13-104c545ffd97"> |

mrcthms replied:
@sam-b-rose No issues from our side with this change if it's the standardised way moving forward.

alex-page commented on polaris-react/src/components/Frame/tests/Frame.test.tsx:
```diff
@@ -58,14 +59,14 @@ describe('<Frame />', () => {
-      expect(skipToContentLinkText).toContainReactComponent('a', {
+      expect(skipToContentLinkText).toContainReactComponent(Text, {
         children: 'Skip to content',
       });
```
Update this to check text inside instead of component

laurkim commented on polaris-react/src/components/ActionList/components/Item/Item.tsx:
```diff
@@ -88,7 +88,9 @@ export function Item({
     </>
   ) : (
-    contentText
+    <Text as="span" variant="bodyMd">
```
```suggestion
    <Text as="span" variant="bodyMd" fontWeight={active ? "semibold" : "regular"}>
```

In the storybook diff for [ActionList with an icon and suffix](https://5d559397bae39100201eedc1-yrjhwxlxrv.chromatic.com/?path=/story/all-components-actionlist--with-an-icon-and-a-suffix), the active item no longer has the bolded font weight ([prod](https://storybook.polaris.shopify.com/?path=/story/all-components-actionlist--with-an-icon-and-a-suffix)).

sam-b-rose replied:
Great catch!

laurkim commented on polaris-react/src/components/Listbox/components/TextOption/TextOption.tsx:
```diff
@@ -35,17 +36,23 @@ export const TextOption = memo(function TextOption({
     <Box width="100%">
       <InlineStack wrap={false} align="space-between" gap="200">
-        {children}
+        {textContent}
```
With the children now wrapped around a single Text component, `InlineStack` no longer applies spacing between each individual child element.
![20-54-lg0k7-a5zwy](https://github.com/Shopify/polaris/assets/26749317/a3c157ba-471a-43e8-91a4-6e24778ddc27)

sam-b-rose replied:
Ah, good catch. This might be a case where we need to rely on dev implementation rather than fix internally. I'll revert this change for now

laurkim commented on polaris-react/src/components/Listbox/components/TextOption/TextOption.tsx:
```diff
@@ -35,17 +36,23 @@ export const TextOption = memo(function TextOption({
   );
 
+  const textContent = (
+    <Text as="span" variant="bodyMd">
```
```suggestion
    <Text as="span" variant="bodyMd" fontWeight={selected ? "semibold" : "regular"}>
```

The `.selected` class font weight isn't being applied for selected TextOptions. The override in the css file can be removed too if we do the above.
![20-00-a537b-2qpgz](https://github.com/Shopify/polaris/assets/26749317/cdecacea-75c6-4960-b523-3bdd7e152392)

laurkim commented on polaris-react/src/components/Navigation/components/Item/Item.tsx:
```diff
@@ -137,16 +140,29 @@ export function Item({
+      ? 'subdued'
+      : undefined;
+  let fontWeight: TextProps['fontWeight'] = 'regular';
+  if (matches && !childIsActive) {
```
```suggestion
  if ((matches && !childIsActive) || (level === 0 && selected)) {
```
Selected L1 items also need `semibold` font weight. Comparing selected Products nav item on [prod](https://storybook.polaris.shopify.com/?path=/story/all-components-navigation--with-an-active-root-item-with-secondary-navigation-items)/[storybook](https://5d559397bae39100201eedc1-yrjhwxlxrv.chromatic.com/?path=/story/all-components-navigation--with-an-active-root-item-with-secondary-navigation-items).

sam-b-rose replied:
Are you sure? When looking at the prod storybook, looks like **Products** is `medium` and the active child item is `semibold`.

<img width="144" alt="21-26-ndob4-e43u2" src="https://github.com/Shopify/polaris/assets/11774595/f7ce5335-50c8-4f73-92e7-e3603648b41b">
<img width="125" alt="21-32-p6wfx-we3uc" src="https://github.com/Shopify/polaris/assets/11774595/f8717a0a-9b2b-48cd-ad31-f3ddade5e95c">

laurkim replied:
Sorry, my terminology was a bit mixed up because we've referred to `level=0` items as L1 for view transition work 😅

There is a discrepancy in the story I linked earlier between this branch's storybook and prod. Home and Products are both active by default and none of the subnavs for Prod are active. On this branch, Home still has `semibold` applied but Products has `medium`.
<img width="300" alt="nav-pr" src="https://github.com/Shopify/polaris/assets/26749317/fc66562e-b1c4-455a-9b65-21df913f7d13">

Whereas in prod, Home has `semibold` because it's active but because Products is active too, it also has `semibold`. 
<img width="300" alt="nav-prod" src="https://github.com/Shopify/polaris/assets/26749317/8543ad46-4592-4598-84c1-9a587da4e9d6">

I actually think the logic in my suggestion can be simplified after testing the story locally and comparing it to prod. I don't think the level matters, we just need to also check that `selected` is true to apply `semibold`
```suggestion
  if ((matches || selected) && !childIsActive) {
```

sam-b-rose replied:
Ah, I see. Good catch, this is an odd state imo where we have two actively selected nav items 🤔 but best to keep the behavior the same. Thanks, Lo!

laurkim commented on polaris-react/src/components/Page/components/Header/components/Title/Title.tsx:
```diff
@@ -33,7 +33,13 @@ export function Title({
-  const titleMarkup = title ? <h1 className={className}>{title}</h1> : null;
+  const titleMarkup = title ? (
+    <h1 className={className}>
+      <Text as="span" variant="headingLg" fontWeight="bold">
```
Just noting that the `headingLg` variant has uses the `--p-font-letter-spacing-dense` as opposed to `normal` so and it's causing a diff in Chromatic but we don't have support for overriding letter spacing on Text so that's probably fine 👍

sam-b-rose replied:
Agreed, I think it is best to lean toward less overrides and follow our text style variant here

laurkim commented on polaris-react/src/components/CalloutCard/CalloutCard.tsx:
```diff
@@ -86,7 +86,9 @@ export function CalloutCard({
                   {title}
```
CalloutCard title font-size is inconsistent with Card/LegacyCard.
The other two use `--p-font-size-325` but CalloutCard has `--p-font-size-350`.
It doesn't come up in the diff for me to add a suggestion but I recommend updating the variant on the Text component wrapping title to `headingSm` instead of `headingMd`.

sam-b-rose replied:
Agreed and done ✅

laurkim commented on polaris-react/src/components/Filters/components/FiltersBar/FiltersBar.tsx:
```diff
@@ -177,7 +173,7 @@ export function FiltersBar({
         }
       >
-        <Text variant={labelVariant} as="span">
+        <Text as="span" variant="bodySm">
```
Without the `color: var(--p-text-color)` styles the color of the Add Filter button is inconsistent with the other filters. 
![20-33-eg1lt-jm6jj](https://github.com/Shopify/polaris/assets/26749317/83eb6c4a-f3f0-45de-ba9c-f51b52b480ec)

sam-b-rose replied:
Good catch. I feel like this shouldn't have to be defined here. Should we be setting the default text color at the root level?

The `Text` component doesn't have a default opinion on color—only is tone is used.

We might want to have a way to specify the default text color on the component here like `tone="neutral"`, `tone="default`", or `tone="base"`

@alex-page @laurkim

laurkim replied:
Just for context, Text has never had a default tone/color because our legacy typography components didn't and we tried to keep that consistent during the typography project 👀

If we want to set it with `tone`, I'm leaning towards `tone="base"` because it's consistent with Icon

alex-page replied:
There are a couple of issues open around `base` but this sounds fine to me.
https://github.com/Shopify/ui-api-design/issues/317
https://github.com/Shopify/ui-api-design/issues/301

laurkim replied:
I didn't want to start a separate thread for this but I noticed the color is also off for disabled filters.
Can we also support `disabled` tone so that we can conditionally set the color as `base` or `disabled` depending on the `disabled` prop @sam-b-rose?
![22-54-hf58l-ps747](https://github.com/Shopify/polaris/assets/26749317/2f99f07d-7b61-49a6-8f34-2413cfd5d0bd)

sam-b-rose replied:
Good catch. I've added a `disabled` option to tone and updated the prop logic for FiltersBar

laurkim reviewed (approved):
Amazing 💯 ⭐ 
A bit of feedback on some UI inconsistencies when reviewing chromatic but looks good overall!

aaronccasanova commented on polaris-react/src/components/AppProvider/global.css:
```diff
@@ -53,18 +53,23 @@
 body,
 button {
   font-family: var(--p-font-family-sans);
+  color: var(--p-color-text);
```
Would the `font-size`/`line-height` updates and added responsiveness/default `button` `color` in the reset be considered a breaking change?

sam-b-rose replied:
Good question. I didn't initially think so, but we could remove the `global.css` changes here just to be safe.

aaronccasanova reviewed (approved):
Awesome work @sam-b-rose!! My review focused on tophatting the new `Text` variants, `Text` styles, `@shopify/stylelint-polaris` rule changes, and Polaris site updates, LGTM 🚀

laurkim commented on polaris-react/src/components/ChoiceList/ChoiceList.tsx:
```diff
@@ -71,12 +72,10 @@ export function ChoiceList({
-    >
-      {title}
+    <Box paddingBlockEnd={{xs: '0', md: '100'}}>
+      <Text as="legend" variant="bodyMd" visuallyHidden={titleHidden}>
```
Looks like `Box` overwrites the default `2px` inline padding when for `legend` elements but since Text has no spacial logic, the default padding is being applied.

Think it might be better to keep Box as `legend` and Text as `span`
Prod
<img width="250" alt="choicelist-prod" src="https://github.com/Shopify/polaris/assets/26749317/a2ccb238-36f3-4fc7-98e3-a49ebf88acc9">
PR
<img width="250" alt="choicelist-pr" src="https://github.com/Shopify/polaris/assets/26749317/0576d60a-75e4-475d-8df2-5d57ed9cbd9f">

laurkim commented on polaris-react/src/components/Choice/Choice.tsx:
```diff
@@ -150,7 +150,9 @@ export function Choice({
       <span className={styles.Control}>{children}</span>
       <span className={styles.Label}>
-        <span>{label}</span>
+        <Text as="span" variant="bodyMd">
```
There's a pretty noticeable difference on font size for small screens stemming from `bodyMd` being responsive.
Just want to double check we're okay with the update style for Choice (and ChoiceList) for xs/sm screens

Prod
<img width="250" alt="choice-prod" src="https://github.com/Shopify/polaris/assets/26749317/3ce87b51-2c82-4b96-a67f-bdb4a965ac40">
PR
<img width="250" alt="choice-pr" src="https://github.com/Shopify/polaris/assets/26749317/bd0a19b6-e4ea-4acb-a237-181383ea7171">

sam-b-rose replied:
@heyjoethomas and I confirmed that this is the desired visual. We are going to adjust the ChoiceList item spacing in a separate PR.

laurkim commented on polaris-react/src/components/DropZone/components/FileUpload/FileUpload.module.css:
```diff
@@ -27,9 +27,6 @@
   text-decoration: none;
-  font-size: var(--p-font-size-300);
-  font-weight: var(--p-font-weight-semibold);
-  line-height: 1;
```
Just want to flag the larger line-height as a result of the `bodySm` styles. 
It's especially noticeable on [smaller screens](http://localhost:6006/?path=/story/all-components-dropzone--default) if you set viewport to `xs`.

sam-b-rose replied:
Good catch. I think this is okay, though. It matches the height and text styles of the default secondary button. Curious why we don't use a primary button here? 🤔 In Figma we are using an actual Button component.

sam-b-rose replied:
I've updated the DropZone (FileUpload) to use a Button. That makes more systematic sense and follows how it is being created in Figma

laurkim commented on polaris-react/postcss-mixins/text-style-input.css:
```diff
@@ -6,6 +6,7 @@
 
   @media (--p-breakpoints-md-up) {
-    font-size: var(--p-font-size-350);
+    font-size: var(--p-font-size-325);
```
Out of curiosity, was the original font size too large?
Even though it's only a 1px difference it's a noticeable change although that may be from me switching very quickly between the 2 snapshots in the [chromatic diff](https://shopify.chromatic.com/test?appId=5d559397bae39100201eedc1&id=65fd8109f568ad30cfc2103a) in Frame.

sam-b-rose replied:
Looking at [the Figma specs](https://www.figma.com/file/jLLkmo9r28hiqPvf4s90E4/Polaris-Gen-3-Components?type=design&node-id=111914-81258&mode=design&t=rO09MT8LxmiYNP3z-4), this should have been `325` (`13px`). Great callout!

laurkim commented on polaris-react/src/components/Filters/Filters.module.css:
```diff
@@ -217,22 +214,12 @@
   margin-left: var(--p-space-200);
-
-  span {
-    font-size: var(--p-font-size-300);
```
Since we're not overriding the font size anymore for the Clear button and it's using the `monochromePlain` prop, it's [sets](https://github.com/Shopify/polaris/blob/880f31b44baa1cfc4a58320f466ccc5c6438bf48/polaris-react/src/components/Button/Button.tsx#L182) the Text variant to `bodyMd` and it looks too large next to all the other filter pills that are using `bodySm`

![index-filters](https://github.com/Shopify/polaris/assets/26749317/99dad6ef-2bca-4cfa-9f06-685878f8853f)

sam-b-rose replied:
I agree. But I also think this is a more of a system issue with how our buttons function with micro + monochrome/plain. This is also the case where we have any plain or monochrome buttons inline with other non-plain buttons

sam-b-rose replied:
@heyjoethomas do you have any thoughts here? Particularly, should we make `<Button variant="monochromePlain | plain" size="micro">` have a `bodySm` text size?

sam-b-rose replied:
Alternatively, monochromePlain and plain could inherit the parent text size. This would help prevent areas where the parent text is not using a set text style:

<img width="296" alt="25-24-bwjx5-55c6p" src="https://github.com/Shopify/polaris/assets/11774595/a1815935-7e9b-4c93-b69c-decdbeeec580">

heyjoethomas replied:
@sam-b-rose Setting it to `bodySm` is probably safe given the intent here is to be small and compact. Otherwise it has the same text style as a default `Plan` button

sam-b-rose replied:
Updated the Button micro with plain/monochrome variant to use the `bodySm` style. This looks much better in the FiltersBar 💯

laurkim reviewed (commented):
Some feedback from tophatting chromatic before most recent force push, I can double check the new chromatic again in a separate review

laurkim commented on polaris-react/src/components/ActionList/components/Item/Item.tsx:
```diff
@@ -103,7 +109,13 @@ export function Item({
-  const textMarkup = <span className={styles.Text}>{contentMarkup}</span>;
+  const textMarkup = (
+    <span className={styles.Text}>
+      <Text as="span" variant="bodyMd">
```
```suggestion
      <Text
        as="span"
        variant="bodyMd"
        fontWeight={active ? 'semibold' : 'regular'}
      >
```
We need a `font-weight` override for `active` items when checking Chromatic diff.
Line 65 in `ActionList.module.css` can be removed where `font-weight` is set to semibold for active items with this suggestion.

![25-22-fhbw3-o076f](https://github.com/Shopify/polaris/assets/26749317/4c1c68fa-5cc8-463e-bba7-08c87e9d6778)

laurkim commented on polaris-react/src/components/Tag/Tag.module.css:
```diff
@@ -1,14 +1,10 @@
-  min-height: var(--pc-tag-height);
-  padding: 0;
   padding-inline: calc(var(--p-space-100) + var(--p-space-050));
+  padding-block: var(--p-space-025);
```
I think this can be removed.
Double checking Figma, Tag should have 20px height. With the changes in this PR, Tag has 22px height, so removing this should resolve the Chromatic diff.

sam-b-rose replied:
Good check. I tried cleaning this up a bit more to remove the `line-height` hidden in the unnecessary `.Text` class. However, this might cause a visual regression for tags with custom content.

I'll keep unchanged for now and remove the added padding. Thanks, Lo!

laurkim reviewed (approved):
Overall re-review looks good! 💯 Thanks for addressing all the earlier feedback 🤝 

I've gone and accepted the chromatic diff changes for things we've discussed (i.e., DropZone button styles, Page header letter spacing, SearchField font size). I left any visual regressions I came across while tophatting or for things still pending discussion (i.e., Monchrome plain buttons).

sam-b-rose committed:
Add more tone options for Text component

sam-b-rose committed:
Applied semantic type styles using the `Text` component

sam-b-rose committed:
Add Stylelint warning for typography property disallow list

sam-b-rose committed:
Add page for typography property disallow list rule

sam-b-rose committed:
Update micro plain/monochrome Button text to small

This fix improves text size alignment on the FilterBar component. The "Clear all" button is now the same size as the other buttons in the FilterBar.
